src/sql/workbench/parts/objectExplorer/browser/iconRenderer.ts

```diff
@@ -0,0 +1,371 @@
+							current.remove();
+							break;
+						}
+					}
+					current = next;
+				}
+			}
+		}
+	}
+}
+
+export class ServerStatusBadge extends BaseBadge {
+
+	private static readonly defaultColorConnected: string = 'rgba(59, 180, 74, 100%)';
+	private static readonly defaultColorDisconnected: string = 'rgba(208, 46, 0, 100%)';
+	private static readonly defaultColorBackground: string = 'rgba(255, 255, 255, 80%)';
+	private colorConnected: string = ServerStatusBadge.defaultColorConnected;
+	private colorDisconnected: string = ServerStatusBadge.defaultColorDisconnected;
+	private colorBackground: string = ServerStatusBadge.defaultColorBackground;
+	public readonly badgeType: string = 'SeverStatusBadge';
+
+	constructor(iconSizePx?: number, topMarginPx?: number, leftMarginPx?: number,
+		colorConnected?: string, colorDisconnected?: string, colorBackground?: string
+	) {
+		super(iconSizePx, topMarginPx, leftMarginPx);
+		this.setColorConnected(colorConnected);
+		this.setColorDisconnected(colorDisconnected);
+		this.setColorBackground(colorBackground);
+	}
+
+	public setColorConnected(color: string): ServerStatusBadge {
+		this.colorConnected = color || this.colorConnected;
+		return this;
+	}
+
+	public setColorDisconnected(color: string): ServerStatusBadge {
+		this.colorDisconnected = color || this.colorDisconnected;
+		return this;
+	}
+
+	public setColorBackground(color: string): ServerStatusBadge {
+		this.colorBackground = color || this.colorBackground;
+		return this;
+	}
+
+	public renderBadge(element: HTMLElement, isConnected: boolean): void {
+		let circleColor: string = isConnected ? this.colorConnected : this.colorDisconnected;
+		let backgroundColor: string = isConnected ? this.colorConnected : this.colorBackground;
+		ServerStatusBadge.renderBadgeInternal(element, circleColor, backgroundColor,
+			this.iconSizePx, this.topMarginPx, this.leftMarginPx);
+	}
+
+	public static renderBadgeInternal(element: HTMLElement, circleColor: string, backgroundColor?: string,
+		iconSizePx?: number, topMarginPx?: number, leftMarginPx?: number
+	) {
+		backgroundColor = backgroundColor || ServerStatusBadge.defaultColorBackground;
+		iconSizePx = iconSizePx || ServerStatusBadge.defaultIconSizePx;
+		topMarginPx = topMarginPx || ServerStatusBadge.defaultTopMarginPx;
+		leftMarginPx = leftMarginPx || ServerStatusBadge.defaultLeftMarginPx;
+
+		let height: string = `${iconSizePx / 16 * 0.25}rem`;
+		let width: string = height;
+		let verticalLocation: string = `${(iconSizePx / 16 * 9) + topMarginPx}px`;
+		let horizontalLocation: string = `${(iconSizePx / 16 * 14) + leftMarginPx}px`;
+		let borderWidth: string = `${iconSizePx / 16 * 0.12}rem`;
+
+		let styleIdPrefix: string = 'ServerStatusBadgeStyle_';
+		let badgeId: string = `id_${generateUuid()}`;
+		let styleId: string = `${styleIdPrefix}${badgeId}`;
+		super.removeExistingBadgeByPrefix(element, styleIdPrefix);
+
+		element.innerHTML = (element.innerHTML || '') +
```

Comment:
Again, I'm worried about essentially spamming the dom with lots of CSS if we use this approach. Looking at the existing codebase, the DecorationsService has patterns to define decorations (such as connected / disconnected in your case) and then render them on an element. Could we reuse this? It seems like we'd put the code you write here in `createCssRule` - see example of
https://github.com/Microsoft/azuredatastudio/blob/88a97a7c5454070540589f3042cb82ccb787156e/src/vs/workbench/services/decorations/browser/decorationsService.ts#L62 - but could follow the same pattern for cleanup, reuse etc. Basically we'd only end up defining the 2 badges and then reference them all in the right places. What do you think?

Reply:
fixed.